openmc/mesh.py

```diff
@@ -1607,83 +1654,107 @@ def write_data_to_vtk(self, filename, datasets, volume_normalization=True):
+            Whether or not to normalize the data by the
+            volume of the mesh elements
+        """
         import vtk
-        from vtk.util import numpy_support as vtk_npsup
+        from vtk.util import numpy_support as nps
 
-        if self.centroids is None:
-            raise RuntimeError("No centroid information is present on this "
-                               "unstructured mesh. Please load this "
-                               "information from a relevant statepoint file.")
+        if self.connectivity is None or self.vertices is None:
+            raise RuntimeError('This mesh has not been '
+                               'loaded from a statepoint file.')
 
-        if self.volumes is None and volume_normalization:
-            raise RuntimeError("No volume data is present on this "
-                               "unstructured mesh. Please load the "
-                               " mesh information from a statepoint file.")
+        if filename is None:
+            filename = f'mesh_{self.id}.vtk'
 
-        # check that the data sets are appropriately sized
-        errmsg = "The size of the dataset {} should be equal to the number of cells"
-        for label, dataset in datasets.items():
-            if isinstance(dataset, np.ndarray):
-                if not dataset.size == self.dimension[0] * self.dimension[1]* self.dimension[2]:
-                    raise RuntimeError(errmsg.format(label))
-            else:
-                if len(dataset) == self.dimension[0] * self.dimension[1]* self.dimension[2]:
-                    raise RuntimeError(errmsg.format(label))
-            cv.check_type('label', label, str)
+        writer = vtk.vtkUnstructuredGridWriter()
 
-        # create data arrays for the cells/points
-        cell_dim = 1
-        vertices = vtk.vtkCellArray()
-        points = vtk.vtkPoints()
+        writer.SetFileName(str(filename))
 
-        for centroid in self.centroids:
-            # create a point for each centroid
-            point_id = points.InsertNextPoint(centroid * self.length_multiplier)
-            # create a cell of type "Vertex" for each point
-            cell_id = vertices.InsertNextCell(cell_dim, (point_id,))
+        grid = vtk.vtkUnstructuredGrid()
+
+        vtk_pnts = vtk.vtkPoints()
+        vtk_pnts.SetData(nps.numpy_to_vtk(self.vertices))
+        grid.SetPoints(vtk_pnts)
+
+        n_skipped = 0
+        elems = []
+        for elem_type, conn in zip(self.element_types, self.connectivity):
+            if elem_type == self._LINEAR_TET:
+                elem = vtk.vtkTetra()
+            elif elem_type == self._LINEAR_HEX:
+                elem = vtk.vtkHexahedron()
+            elif elem_type == self._UNSUPPORTED_ELEM:
+                n_skipped += 1
+            else:
+                raise RuntimeError(f'Invalid element type {elem_type} found')
+            for i, c in enumerate(conn):
+                if c == -1:
+                    break
```

Comment:
Should this really be `break` and not `continue`?

Reply:
It really should be `break`. Each element has connectivity with 8 entries, so if we're working with a linear tetrahedron we'll stop adding points once we reach that first `-1` value.